### PR TITLE
Client-side logic for writing to the database.

### PR DESCRIPTION
**Pull request by navilperez**
index.html: Basic UI for the Blockly workspace.

index.js: Entry point for the client. Includes a change listener that
detects events created on the UI and logic to send those changes to the
server.

WorkspaceClient.js: Class for managing events and handling event
resolution. This change adds support for tracking events created on the
client and writing them to the database.

### CODE REVIEW

File: blockly-rtc/public/WorkspaceClient.js [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        this.beginWrite_();
+        return await writeEvents(this.inProgress)
+        .then(() => {
+            this.endWrite_(true);
+        })
+        .catch(() => {
+            this.endWrite_(false);
+            throw 'Failed to write to database.';
+        });
+    };
```
Comment (samelhusseini) [inline blockly-rtc/public/WorkspaceClient.js]: There's a mix of async await here and promises. You can instead do: 
```
this.beginWrite_();
try {
    const events = await writeEvents(this.inProgress);
    this.endWrite_(true);
    return events;
} catch {
   this.endWrite_(false);
   throw Error('Failed to write to database.');
}
```

File: blockly-rtc/public/index.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            media: 'media/'
+        });
+    var workspaceClient = new WorkspaceClient(workspace.id);
+
+    workspace.addChangeListener(function(event) {
+        if (event instanceof Blockly.Events.Ui) {
+            return;
+        };
+        workspaceClient.addEvent(event.toJson());
+        if (! Blockly.Events.getGroup()) {
```
Comment (samelhusseini) [inline blockly-rtc/public/index.js]: nit: remove extra space after ``!``

File: blockly-rtc/public/index.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        };
+        if (workspaceClient.notSent.length == 0) {
+            return;
+        };
+        workspaceClient.writeToDatabase()
+        .then(() => {
+            sendChanges();
+        })
+        .catch(() => {
+            console.error();
```
Comment (samelhusseini) [inline blockly-rtc/public/index.js]: Nothing is passed to the console.error..

File: blockly-rtc/public/WorkspaceClient.js (hunk 1 above)
Comment (navilperez) [reply blockly-rtc/public/WorkspaceClient.js]: Thanks for the suggestion. I prefer to keep it consistent so this is great :)

File: blockly-rtc/public/WorkspaceClient.js [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+ * @author navil@google.com (Navil Perez)
+ */
+
+import {getEvents, writeEvents} from './api';
+
+/**
+ * Class for managing events between the workspace and the server.
+ * @param {string} workspaceId The id of the Blockly.Workspace instance this
+ * client corresponds to.
+ * */
```
Comment (samelhusseini) [inline blockly-rtc/public/WorkspaceClient.js]: That's a broken jsdoc. It should end with a `*/` on its own line.

File: blockly-rtc/public/index.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+/**
+ * @fileoverview Realtime collaboration demo using Node.js and SQlite.
+ * @author navil@google.com (Navil Perez)
+ */
+
+import * as Blockly from 'blockly';
+import WorkspaceClient from './WorkspaceClient';
+
+document.addEventListener("DOMContentLoaded", () => {
```
Comment (samelhusseini) [inline blockly-rtc/public/index.js]: nit: single quote

File: blockly-rtc/test/test.js [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Unit tests for WorkspaceClient.
+ * @author navil@google.com (Navil Perez)
+ */
+
+const assert = require('assert');
+const sinon = require('sinon');
```
Comment (samelhusseini) [inline blockly-rtc/test/test.js]: Did you add sinon to ``package.json``?

File: blockly-rtc/test/test.js [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+const api = require('../public/api');
+const WorkspaceClient = require('../public/WorkspaceClient').default;
+
+suite('WorkspaceClient', () => {
+
+  suite('writeToDatabase()', () => {
+    test('Write succeeds, events move from notSent to inProgress.', async () => {
+      const writeEventsStub = sinon.stub(api, "writeEvents").resolves();
```
Comment (samelhusseini) [inline blockly-rtc/test/test.js]: nit: single quote.

File: blockly-rtc/public/WorkspaceClient.js [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        this.writeInProgress = false;
+        this.counter = 0;
+    };
+
+    /**
+     * Create and entry from an event and add it to active changes.
+     * An entry is of the form {"event": Blockly.Event, "entryId": string} where
+     * event is an event created on the workspace and entryId is of the form
+     * {workspaceId}:{counter}.
+     * @param {boolean} success Indicates the success of the database write.
```
Comment (samelhusseini) [inline blockly-rtc/public/WorkspaceClient.js]: The param here is not the same name, nor the right type?

File: blockly-rtc/public/WorkspaceClient.js [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        this.counter += 1;
+        this.activeChanges.push({
+            event: event,
+            entryId: entryId
+        });
+    };
+
+    /**
+     * Add all events from a Blockly.Events group to notSent.
+     * @public
```
Comment (samelhusseini) [inline blockly-rtc/public/WorkspaceClient.js]: What's the purpose of this. Is it to flush pending changes? If so, perhaps flush is a better name.

File: blockly-rtc/public/WorkspaceClient.js (hunk 8 above)
Comment (navilperez) [reply blockly-rtc/public/WorkspaceClient.js]: Correct. Thanks for noting that.

File: blockly-rtc/public/WorkspaceClient.js (hunk 9 above)
Comment (navilperez) [reply blockly-rtc/public/WorkspaceClient.js]: Yes, flush pending changes and move those changes to the notSent array. I am okay with the name change.

File: blockly-rtc/test/test.js (hunk 6 above)
Comment (navilperez) [reply blockly-rtc/test/test.js]: Yes I have. I have now uploaded package.json to this change.

File: blockly-rtc/webpack.config.js [hunk 10]
```diff
@@ -18,7 +18,7 @@
 
 /**
  * @fileoverview Webpack configuration file.
- * @author samelh@google.com (Sam El-Husseini)
+ * @author navil@google.com (Sam El-Husseini)
```
Comment (navilperez) [inline blockly-rtc/webpack.config.js]: Realizing I did not change the author name. I will upload the fix in a following commit along with any other changes requested.

**Comment by rachel-fenichel** [comment]
I'm late to the party, but this looks pretty good once you clean up what Sam pointed out.

File: blockly-rtc/src/WorkspaceClient.js [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        this.writeInProgress = false;
+        this.counter = 0;
+    };
+
+    /**
+     * Create and entry from an event and add it to active changes.
+     * An entry is of the form {"event": Blockly.Event, "entryId": string} where
+     * event is an event created on the workspace and entryId is of the form
+     * {workspaceId}:{counter}.
+     * @param {!Blockly.Events} event The Blockly.Event created by the client.
```
Comment (samelhusseini) [inline blockly-rtc/src/WorkspaceClient.js]: The type here should be ``!Blockly.Events.Abstract``.

Comment (samelhusseini) [reply blockly-rtc/src/WorkspaceClient.js]: Actually you're only passing the event.toJson() product to this method, so it's ``!Object``

File: package-lock.json [hunk 12]
```diff
@@ -0,0 +1,691 @@
+{
```
Comment (samelhusseini) [inline package-lock.json]: This file is unrelated to your change. You can leave this one out.

Comment (samelhusseini) [reply package-lock.json]: This is the top-level package-lock.json

**Review by samelhusseini — COMMENTED** [review]
A few more comments, but otherwise looks good.

**Review by samelhusseini — APPROVED** [review]
